software/create_custom_firmware_uf2.md

```diff
@@ -0,0 +1,117 @@
+# Creating a custom firmware.uf2 image for EuroPi
+
+## Compiling the firmware
+```
+# install extra tools if not already done so
+$ sudo apt update
+$ sudo apt install cmake gcc-arm-none-eabi libnewlib-arm-none-eabi build-essential
+
+# get micropython source from github repo
+# choose latest stable version instead of the master branch
+$ git clone -b v1.19.1 https://github.com/micropython/micropython.git
+
+# fetch submodules
+$ cd micropython
+$ make -C ports/rp2 submodules
+
+# build tools
+$ make -C mpy-cross
+
+# enter the dir for our hardware port
+$ cd ports/rp2
+
+$ ls
+# files that should be frozen and included in the firmware should be placed in the 'modules' dir
+# so thats 'micropython/ports/rp2/modules/'
+# what should be done here is written below
+# once that has been done you can continue to make the firmware
+
+# optional clean up
+$ make clean
+
+# build the firmware
+$ make
+```
+
+## Modules dir
+after the files for EuroPi, EuroPi contrib and ssd1306 are inserted
+the structure of the modules dir should look like this:
+```
+micropython/ports/rp2/modules/
+├── _boot_fat.py
+├── bootloader.py
+├── _boot.py
+├── calibrate.py
+├── contrib
+│   ├── bernoulli_gates.py
+│   ├── coin_toss.py
+│   ├── consequencer.py
+│   ├── cvecorder.py
+│   ├── diagnostic.py
+│   ├── hamlet.py
+│   ├── harmonic_lfos.py
+│   ├── hello_world.py
+│   ├── master_clock.py
+│   ├── menu.py
+│   ├── noddy_holder.py
+│   ├── polyrhythmic_sequencer.py
+│   ├── poly_square.py
+│   ├── probapoly.py
+│   ├── radio_scanner.py
+│   ├── scope.py
+│   ├── smooth_random_voltages.py
+│   ├── strange_attractor.py
+│   └── turing_machine.py
```

Comment:
[optional]: tidy this up by omitting the full list since it changes occasionally.
```suggestion
│   └── ...
```